services/supervisord/supervisord.go

```diff
@@ -157,87 +204,145 @@ func (s *Service) supervisorctl(args ...string) ([]byte, error) {
 	if err != nil {
-		return errors.WithStack(err)
+		return 0, errors.WithStack(err)
 	}
 	if err = p.Signal(unix.SIGUSR2); err != nil {
-		s.l.Warn(err)
+		s.l.Warnf("Failed to send SIGUSR2: %s", err)
 	}
+	s.l.Debug("Waiting for logreopen...")
 	<-ch
 
-	// check log file size for debugging
+	var offset uint32
 	fi, err := os.Stat(pmmUpdatePerformLog)
-	if err != nil {
+	switch {
+	case err == nil:
+		if fi.Size() != 0 {
+			s.l.Warnf("Unexpected log file size: %+v", fi)
+		}
+		offset = uint32(fi.Size())
+	case os.IsNotExist(err):
+		// that's expected as we remove this file above
+	default:
 		s.l.Warn(err)
 	}
-	if fi.Size() != 0 {
-		s.l.Warnf("%+v", fi)
-	}
 
 	_, err = s.supervisorctl("start", pmmUpdatePerformProgram)
-	return err
+	return offset, err
 }
 
-// PMMUpdateRunning returns true if pmm-update-perform supervisord program is running or being restarted,
+// UpdateRunning returns true if pmm-update-perform supervisord program is running or being restarted,
 // false if it is not running / failed.
-func (s *Service) PMMUpdateRunning() bool {
-	s.rw.RLock()
-	defer s.rw.RUnlock()
-
-	// first check with status command is case we missed that event during maintail or pmm-managed restart
+func (s *Service) UpdateRunning() bool {
+	// First check with status command is case we missed that event during maintail or pmm-managed restart.
+	// See http://supervisord.org/subprocess.html#process-states
 	b, err := s.supervisorctl("status", pmmUpdatePerformProgram)
 	if err != nil {
 		s.l.Warn(err)
 	}
 	s.l.Debugf("%s", b)
 	if f := strings.Fields(string(b)); len(f) > 2 {
-		switch f[1] {
-		case "FATAL":
+		switch status := f[1]; status {
+		case "FATAL", "STOPPED": // will not be restarted
 			return false
 		case "STARTING", "RUNNING", "BACKOFF", "STOPPING":
 			return true
+		case "EXITED":
+			// it might be restarted - we need to inspect last event
 		default:
-			// we need to inspect last event
+			s.l.Warnf("Unknown %s process status %q.", pmmUpdatePerformProgram, status)
+			// inspect last event
 		}
 	}
 
-	switch s.pmmUpdatePerformLastEvent {
-	case starting, running, exitedUnexpected:
+	s.eventsM.Lock()
+	lastEvent := s.pmmUpdatePerformLastEvent
+	s.eventsM.Unlock()
+
+	switch lastEvent {
+	case stopping, starting, running:
+		return true
+	case exitedUnexpected: // will be restarted
 		return true
-	case fatal:
+	case exitedExpected, fatal: // will not be restarted
 		return false
+	case stopped: // we don't know
+		fallthrough
 	default:
-		s.l.Warnf("Unknown %s status.", pmmUpdatePerformProgram)
+		s.l.Warnf("Unhandled %s status (last event %q), assuming it is not running.", pmmUpdatePerformProgram, lastEvent)
 		return false
 	}
 }
+
+// UpdateLog returns some lines and a new offset from pmm-update-perform log starting from the given offset.
+// It may return zero lines and the same offset. Caller is expected to handle this.
+func (s *Service) UpdateLog(offset uint32) ([]string, uint32, error) {
```

Comment:
Maybe working with log should be extracted from supervisord?

Reply:
Yeah, that's a tricky one, I spent quite a lot of time moving it around. In the end, I think it belongs there (for now) because:
* It is tightly integrated with `StartUpdate` which uses supervisord in several ways.
* supervisord is currently the only package that requires special and new (devcontainer) treatment for testing. Until we talk about PMM Server development setup in general (hopefully in Zagreb), I would like to keep in contained.